stable_baselines3/common/vec_env/vec_frame_stack.py

```diff
@@ -28,68 +28,166 @@ def __init__(self, venv: VecEnv, n_stack: int, channels_order: Optional[str] = N
-            if is_image_space(wrapped_obs_space):
-                self.channels_first = is_image_space_channels_first(wrapped_obs_space)
+            if is_image_space(obs_space):
+                channels_first = is_image_space_channels_first(obs_space)
             else:
                 # Default behavior for non-image space, stack on the last axis
-                self.channels_first = False
+                channels_first = False
         else:
-            assert channels_order in {"last", "first"}, "`channels_order` must be one of following: 'last', 'first'"
+            assert channels_order in {
+                "last",
+                "first",
+            }, "`channels_order` must be one of following: 'last', 'first'"
 
-            self.channels_first = channels_order == "first"
+            channels_first = channels_order == "first"
 
         # This includes the vec-env dimension (first)
-        self.stack_dimension = 1 if self.channels_first else -1
-        repeat_axis = 0 if self.channels_first else -1
-        low = np.repeat(wrapped_obs_space.low, self.n_stack, axis=repeat_axis)
-        high = np.repeat(wrapped_obs_space.high, self.n_stack, axis=repeat_axis)
-        self.stackedobs = np.zeros((venv.num_envs,) + low.shape, low.dtype)
-        observation_space = spaces.Box(low=low, high=high, dtype=venv.observation_space.dtype)
-        VecEnvWrapper.__init__(self, venv, observation_space=observation_space)
+        stack_dimension = 1 if channels_first else -1
+        repeat_axis = 0 if channels_first else -1
+        low = np.repeat(obs_space.low, self.n_stack, axis=repeat_axis)
+        high = np.repeat(obs_space.high, self.n_stack, axis=repeat_axis)
+        stackedobs = np.zeros((self.venv.num_envs,) + low.shape, low.dtype)
+        observation_space = spaces.Box(low=low, high=high, dtype=obs_space.dtype)
+        return channels_first, stack_dimension, stackedobs, observation_space
+
+    def step_wait(
+        self,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, List[Dict[str, Any]]]:
 
-    def step_wait(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, List[Dict[str, Any]]]:
         observations, rewards, dones, infos = self.venv.step_wait()
-        # Let pytype know that observation is not a dict
-        assert isinstance(observations, np.ndarray)
-        stack_ax_size = observations.shape[self.stack_dimension]
-        self.stackedobs = np.roll(self.stackedobs, shift=-stack_ax_size, axis=self.stack_dimension)
-        for i, done in enumerate(dones):
-            if done:
-                if "terminal_observation" in infos[i]:
-                    old_terminal = infos[i]["terminal_observation"]
-                    if self.channels_first:
-                        new_terminal = np.concatenate(
-                            (self.stackedobs[i, :-stack_ax_size, ...], old_terminal), axis=self.stack_dimension
-                        )
+
+        if isinstance(self.venv.observation_space, spaces.Box):
```

Comment:
Again, looks like a lot of branching... maybe it would also make sense to create a separate wrapper for clarity?

Reply:
Although I agree its a lot of branching, if there was a separate wrapper would it mean that we'd have to keep track of whether the environment is a `VecFrameStack` as opposed to a `DictVecFrameStack`?.

Additionally, from a usability point of view, assuming that Dicts are supported "natively" because the correct Buffer class is chosen based on the observation space (e.g., `DictReplayBuffer`), I would expect that `VecFrameStack` to have similar functionality so as to avoid:
```
env = ...
if isinstance(env.observation_space, gym.spaces.Dict):
    env = DictVecFrameStack(env, n_stack=4)
else:
    env = VecFrameStack(env, n_stack=4)
```

Is there interest in an alternative that doesn't require an another wrapper class?

Potentially there could be a class that holds the data of `VecFrameStack.stacked_obs` and `VecFrameStack` can call methods upon the data. Thoughts?

Reply:
Good point.
Maybe as an intermediate solution we could have private implementations for each case and `VecFrameStack` just do:

```python
class VecFrameStack(VecEnvWrapper):
    def __init__(self, venv: VecEnv, n_stack: int, channels_order: Optional[str] = None):
        if isinstance(venv.observation_space, spaces.Box):
            return VecFrameStackBox(venv, n_stack, channels_order)

        elif isinstance(wrapped_obs_space, spaces.Dict):
            return VecFrameStackDict(venv, n_stack, channels_order)
        else:
            raise Exception("VecFrameStack only works with gym.spaces.Box and gym.spaces.Dict observation spaces")
```

Reply:
btw, do we want to stack every key (then we would need to check that each key is a Box) or allow to stack only some keys?

Reply:
> we could have private implementations for each case ...

I don't think returning the private implementations like this will work. I think there are (and if not there could be) checks to see if the environment has a `VecFrameStack` wrapping. In this case, both `VecFrameStackBox` and `VecFrameStackDict` would have to be child classes of `VecFrameStack` which is confusing if their parent `VecFrameStack` depends upon them. Additionally the `__init__` method does not return anything so this would cause a `TypeError`. 

As an alternative, I was imagining having a private class of `StackedObservations` that wraps the data. Something like:

```
class StackedObservations:
    def __init__(self, num_stacks, observation_space, channels_order):
        ...

    def reset(self, observation):
        ...
        return self.stackedobs

    def update(self, observations, dones, info):
        ....
        return self.stackedobs, info

class StackedDictObservations(StackedObservations):
    ...
```

This would replace `compute_stacking` completely avoiding code duplication and enabling reuse `StackedObservations` outside of  `VecFrameStack` and could be used as:

```
class VecFrameStack(VecEnvWrapper):
    def __init__(self, venv: VecEnv, n_stack: int, channels_order: Optional[str] = None):
        if isinstance(venv.observation_space, spaces.Box):
            self.stacked_observations= StackedObservations(num_stacks, observation_space, channels_order)
        elif isinstance(venv.observation_space, spaces.Dict):
            self.stacked_observations= StackedDictObservations(num_stacks, observation_space, channels_order)
        else:
            raise Exception("VecFrameStack only works with gym.spaces.Box and gym.spaces.Dict observation spaces")

    def step_wait(self):
        observations, rewards, dones, infos = self.venv.step_wait()
        stacked_obs, infos = self.stacked_observations.update(observations, dones, infos)
        return stacked_obs, rewards, dones, infos

    def reset(self):
        observation = self.venv.reset()
        return self.stacked_observations.reset(observation)
```

In the future, this allows for custom stacking techniques but I do think that by default we should stack every key. Generic implementations are more easily customized than selective ones IMO. 

Thoughts?

Reply:
>In the future, this allows for custom stacking techniques but I do think that by default we should stack every key. Generic implementations are more easily customized than selective ones IMO.
>Thoughts?

>As an alternative, I was imagining having a private class of StackedObservations that wraps the data. Something like:

Looks like a good compromise ;)

Reply:
Deal ;)

Reply:
Put this in the latest commit. I still use `compute_stacking` within the new `StackedObservation` class to reduce code duplication. I think it really cleans up `VecFrameStack` . Perhaps one can pass in their own kind of stacking class to `VecFrameStack` but I don't really want the PR to get even bigger than it is so this can be done in the future if desired.